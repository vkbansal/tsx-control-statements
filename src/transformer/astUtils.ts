import ts from 'typescript';

export function getChildrenWrappedInFragmentElement(
	context: ts.TransformationContext,
	node: ts.JsxChild[],
	fragment = 'React.Fragment',
): ts.JsxElement | ts.NullLiteral {
	return node.length > 0
		? context.factory.createJsxElement(
				context.factory.createJsxOpeningElement(
					context.factory.createIdentifier(fragment),
					[],
					context.factory.createJsxAttributes([]),
				),
				node,
				context.factory.createJsxClosingElement(context.factory.createIdentifier(fragment)),
		  )
		: context.factory.createNull();
}

export function createNestedTerinaryExpression(
	this: ts.SourceFile,
	context: ts.TransformationContext,
	nodes: ts.JsxElement[],
	fallback: ts.JsxElement | null,
	visitor: ts.Visitor,
): ts.Expression {
	const [node, ...restNodes] = nodes;

	const condition = node.openingElement.attributes.properties.find((attribute) => {
		return attribute.name?.getText(this) === 'condition';
	});

	if (
		condition &&
		ts.isJsxAttribute(condition) &&
		condition.initializer &&
		ts.isJsxExpression(condition.initializer) &&
		condition.initializer.expression
	) {
		const children = node.children.map(visitor.bind(this)) as ts.JsxChild[];
		const whenTrue = context.factory.createParenthesizedExpression(
			getChildrenWrappedInFragmentElement(context, children),
		);
		const whenFalse =
			restNodes.length > 0
				? context.factory.createParenthesizedExpression(
						createNestedTerinaryExpression.call(this, context, restNodes, fallback, visitor),
				  )
				: fallback
				? context.factory.createParenthesizedExpression(
						getChildrenWrappedInFragmentElement(
							context,
							fallback.children.map(visitor.bind(this)) as ts.JsxChild[],
						),
				  )
				: context.factory.createNull();

		return context.factory.createConditionalExpression(
			condition.initializer.expression,
			undefined,
			whenTrue,
			undefined,
			whenFalse,
		);
	}

	return context.factory.createNull();
}
