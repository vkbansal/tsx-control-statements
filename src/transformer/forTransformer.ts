import ts from 'typescript';

export default function forTransformer(
	this: ts.SourceFile,
	context: ts.TransformationContext,
	node: ts.JsxElement,
	visitor: ts.Visitor,
): ts.Node {
	const items = node.openingElement.attributes.properties.find((attribute) => {
		return attribute.name?.getText(this) === 'items';
	});

	const children = node.children.filter(
		(child) =>
			ts.isJsxExpression(child) &&
			child.expression &&
			(ts.isArrowFunction(child.expression) || ts.isFunctionExpression(child.expression)),
	) as ts.JsxExpression[];

	const firstChild = children[0];

	/**
	 * Check if first-child is a JSXExpression and the expression is defined
	 */
	if (
		items &&
		ts.isJsxAttribute(items) &&
		items.initializer &&
		ts.isJsxExpression(items.initializer) &&
		items.initializer.expression &&
		firstChild &&
		firstChild.expression
	) {
		const { expression } = firstChild;
		let newExpression: ts.Expression = expression;

		// if expression is an arrow function
		if (ts.isArrowFunction(expression)) {
			let body: ts.Expression | ts.Block | ts.EmptyStatement = ts.factory.createEmptyStatement();

			if (ts.isJsxElement(expression.body)) {
				// if body is JSXElement

				// () => <div/>
				body = visitor.call(this, expression.body) as ts.Expression;
			} else if (ts.isBlock(expression.body)) {
				body = context.factory.createBlock(
					// if body is Block
					// () => {
					//  return <div/>
					// }

					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
					expression.body.statements.map(visitor.bind(this)!) as ts.Statement[],
				);
			} else if (ts.isParenthesizedExpression(expression.body)) {
				// if body is ParenthesizedExpression
				// () => (<div/>)

				// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
				body = expression.body.forEachChild(visitor.bind(this)!) as ts.ConciseBody;
			} else {
				body = expression.body;
			}

			newExpression = context.factory.createArrowFunction(
				expression.modifiers,
				expression.typeParameters,
				expression.parameters,
				expression.type,
				expression.equalsGreaterThanToken,
				body,
			);
		}

		// if expression is an regular function
		if (ts.isFunctionExpression(expression)) {
			newExpression = context.factory.createFunctionExpression(
				expression.modifiers,
				expression.asteriskToken,
				expression.name,
				expression.typeParameters,
				expression.parameters,
				expression.type,
				context.factory.createBlock(
					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
					expression.body.statements.map(visitor.bind(this)!) as ts.Statement[],
				),
			);
		}

		const func = context.factory.createCallExpression(
			context.factory.createPropertyAccessExpression(items.initializer.expression, 'map'),
			undefined,
			[newExpression],
		);

		return node.parent && ts.isJsxElement(node.parent)
			? context.factory.createJsxExpression(undefined, func)
			: func;
	}

	return node;
}
