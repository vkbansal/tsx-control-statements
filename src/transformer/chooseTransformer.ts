import ts from 'typescript';

import { createNestedTerinaryExpression } from './astUtils';

export default function chooseTransformer(
	this: ts.SourceFile,
	context: ts.TransformationContext,
	node: ts.JsxElement,
	visitor: ts.Visitor,
): ts.Node {
	const whenElements: ts.JsxElement[] = [];
	let otherwiseElement: ts.JsxElement | null = null;

	for (const child of node.children) {
		if (!ts.isJsxElement(child)) {
			continue;
		}

		const tagName = child.openingElement.tagName.getText(this);

		if (tagName === 'Otherwise') {
			otherwiseElement = child;
			break;
		}

		if (tagName === 'When') {
			whenElements.push(child);
		}
	}

	if (whenElements.length > 0) {
		const terinary = createNestedTerinaryExpression.call(
			this,
			context,
			whenElements,
			otherwiseElement,
			visitor,
		);

		return node.parent && ts.isJsxElement(node.parent)
			? context.factory.createJsxExpression(undefined, terinary)
			: terinary;
	}

	return node;
}
