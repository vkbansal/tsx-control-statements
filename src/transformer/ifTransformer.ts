import ts from 'typescript';

import { createNestedTerinaryExpression } from './astUtils';

export default function ifTransformer(
  this: ts.SourceFile,
  node: ts.JsxElement,
  visitor: ts.Visitor
) {
  const terinary = createNestedTerinaryExpression.call(this, [node], null, visitor);

  return node.parent && ts.isJsxElement(node.parent)
    ? ts.createJsxExpression(undefined, terinary)
    : terinary;
}
