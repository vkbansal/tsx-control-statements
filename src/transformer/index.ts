import ts from 'typescript';

import ifTransformer from './ifTransformer';
import chooseTransformer from './chooseTransformer';
import forTransformer from './forTransformer';

const IMPORT_REGEX = /@vkbansal\/tsx-control-statements/;

export default function transformer(
  context: ts.TransformationContext
): ts.Transformer<ts.SourceFile> {
  let isImported = false;

  function visitor(this: ts.SourceFile, node: ts.Node): ts.Node {
    if (ts.isImportDeclaration(node) && IMPORT_REGEX.test(node.moduleSpecifier.getText(this))) {
      isImported = true;
      return ts.createEmptyStatement();
    }

    if (ts.isJsxElement(node)) {
      switch (node.openingElement.tagName.getText(this)) {
        case 'If':
          return isImported ? ifTransformer.call(this, node, visitor.bind(this)) : node;
        case 'Choose':
          return isImported ? chooseTransformer.call(this, node, visitor.bind(this)) : node;
        case 'For':
          return isImported ? forTransformer.call(this, node, visitor.bind(this)) : node;
      }
    }

    return ts.visitEachChild(node, visitor.bind(this), context);
  }

  return (sourceFile: ts.SourceFile) => ts.visitNode(sourceFile, visitor.bind(sourceFile));
}
