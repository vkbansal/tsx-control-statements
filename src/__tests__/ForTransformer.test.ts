import * as fs from 'fs';
import * as path from 'path';

import transformer from '../transformer';
import { serializer, getTransformedCode } from './helpers';

expect.addSnapshotSerializer(serializer);

describe('<If /> tests', () => {
	test('transforms <If />', async () => {
		const filename = 'ForStatement.tsx';
		const content = await fs.promises.readFile(
			path.resolve(__dirname, 'fixtures', filename),
			'utf-8',
		);

		const code = await getTransformedCode(content, filename, [transformer]);

		expect(code).toMatchInlineSnapshot(`
      File: ForStatement.tsx

      Content:

      import React from 'react';

      import { For, If } from '@vkbansal/tsx-control-statements';

      export function ForStatement(): React.ReactElement {
        return (
          <div>
            <For items={[1, 2, 3, 4, 5]}>
              {(item) => (
                <If condition={item > 3}>
                  <div key={item}>{item}</div>
                </If>
              )}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {(item) => {
                return (
                  <If condition={item > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {(item) => {
                const a = item;

                return (
                  <If condition={a > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {function (item) {
                return (
                  <If condition={item > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {function (item) {
                const a = item;

                return (
                  <If condition={a > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2]}>{(k) => <div>{k}</div>}</For>
          </div>
        );
      }


      Code before Transform:

      import React from 'react';
      import { For, If } from '@vkbansal/tsx-control-statements';
      export function ForStatement(): React.ReactElement {
        return (
          <div>
            <For items={[1, 2, 3, 4, 5]}>
              {(item) => (
                <If condition={item > 3}>
                  <div key={item}>{item}</div>
                </If>
              )}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {(item) => {
                return (
                  <If condition={item > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {(item) => {
                const a = item;
                return (
                  <If condition={a > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {function (item) {
                return (
                  <If condition={item > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2, 3, 4, 5]}>
              {function (item) {
                const a = item;
                return (
                  <If condition={a > 3}>
                    <div key={item}>{item}</div>
                  </If>
                );
              }}
            </For>
            <For items={[1, 2]}>{(k) => <div>{k}</div>}</For>
          </div>
        );
      }


      Code after Transform: 

      import React from 'react';
      export function ForStatement(): React.ReactElement {
        return (
          <div>
            {[1, 2, 3, 4, 5].map((item) =>
              item > 3 ? (
                <React.Fragment>
                  <div key={item}>{item}</div>
                </React.Fragment>
              ) : null
            )}
            {[1, 2, 3, 4, 5].map((item) => {
              return item > 3 ? (
                <React.Fragment>
                  <div key={item}>{item}</div>
                </React.Fragment>
              ) : null;
            })}
            {[1, 2, 3, 4, 5].map((item) => {
              const a = item;
              return a > 3 ? (
                <React.Fragment>
                  <div key={item}>{item}</div>
                </React.Fragment>
              ) : null;
            })}
            {[1, 2, 3, 4, 5].map(function (item) {
              return item > 3 ? (
                <React.Fragment>
                  <div key={item}>{item}</div>
                </React.Fragment>
              ) : null;
            })}
            {[1, 2, 3, 4, 5].map(function (item) {
              const a = item;
              return a > 3 ? (
                <React.Fragment>
                  <div key={item}>{item}</div>
                </React.Fragment>
              ) : null;
            })}
            {[1, 2].map((k) => (
              <div>{k}</div>
            ))}
          </div>
        );
      }

    `);
	});
});
