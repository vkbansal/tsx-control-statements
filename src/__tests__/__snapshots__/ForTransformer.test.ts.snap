// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<If /> tests transforms <If /> 1`] = `
File: ForStatement.tsx

Content:

import React from 'react';

import { For } from '@vkbansal/tsx-control-statements';

export function ForStatement(): React.ReactElement {
  return (
    <div>
      <For items={[1, 2, 3, 4, 5]}>{(item) => <div key={item}>{item}</div>}</For>
      <For items={[1, 2, 3, 4, 5]}>
        {function(item) {
          return <div key={item}>{item}</div>;
        }}
      </For>
    </div>
  );
}


Code before Transform:

import React from 'react';
import { For } from '@vkbansal/tsx-control-statements';
export function ForStatement(): React.ReactElement {
  return (
    <div>
      <For items={[1, 2, 3, 4, 5]}>{(item) => <div key={item}>{item}</div>}</For>
      <For items={[1, 2, 3, 4, 5]}>
        {function(item) {
          return <div key={item}>{item}</div>;
        }}
      </For>
    </div>
  );
}


Code after Transform: 

import React from 'react';
export function ForStatement(): React.ReactElement {
  return (
    <div>
      {[1, 2, 3, 4, 5].map((item) => (
        <div key={item}>{item}</div>
      ))}
      {[1, 2, 3, 4, 5].map(function(item) {
        return <div key={item}>{item}</div>;
      })}
    </div>
  );
}

`;
